### DISCUSSION TITLE
Airflow on Openshift

### DISCUSSION THREAD

**Discussion by yogeshwariingale** [apache/airflow#30322] [Q&A]
Hi,

I am trying to deploy apache airflow on Openshift 4. I am able to deploy flow. But Every Dag is failing. Steps i followed for Deployment
1. I have created template in openshift from below git repository
https://github.com/CSCfi/airflow-openshift

2. Then I installed that Template in openshift. It was giving error with securityContext and runAsUser for all deployment yamls, So I removed these parameters from Deployment.yaml
3. Then It worked. but when i am trying to run any DAG i get below error
*** Log file does not exist: /opt/airflow/logs/get_data/make_tmp_folder/2023-03-27T08:35:25.074418+00:00/3.log
*** Fetching from: http://:8793/log/get_data/make_tmp_folder/2023-03-27T08:35:25.074418+00:00/3.log
*** Failed to fetch log file from worker. The request to ':///' is missing either an 'http://'                         or 'https://' protocol.

Can anyone please provide detail guide to deploy Airflow on Openshift4 with kubernetesExecutor

**Comment by potiuk** [comment]
I tihnk it's best to contact the author of the template and raise your issue there if you need help that is specific to the template. 

I would personally advise you to use our Helm Chart - https://airflow.apache.org/docs/helm-chart/stable/index.html as then it is more likely someone here might help you with the problms you have (for example there runAsUser and security context are configurable so you can adjust them to the OpenShift specific configuration you have (as far as I know OpenShift is very configurable there and your OpenShift might have some specific requirements there for who the users are and whwere they are. There are many problems the Helm Chart solves that you will hit (and have to solve yourself) if you take a random deployment published by someone. From a brief look at the deployment it has plenty of bad practices that we absolutely discourage (for example it tries to dynamicalliy install requirements in your images instead of you having to build your own image upfront if you need more requirements). 

You can see our helm chart "Production Guide" documentation for this for example and many other problems are addressed in the configurable Helm chart of ours.

The specfic problem you have might (but this is a wild guess) come with specific configuration of the IP address/hostname that webserver can communicate with. Webserver needs to communicate with the worker and if the networking rules prevent it, or if your worker cannot retrieve by default the public name of the worker it runs with, this might casue such an issue.

For example if your OpenShift does not provide reverse-dns service that can tell the worker  what it's name is, you might have exactly this problem as you described - you might take a look at  https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#hostname-callable  and revert to IP address as explained there.

**Reply by yogeshwariingale** [reply]
Hi,

I tried to deploy airflow helm chart into Openshift, I am getting below error. I am using gitops and deploying official helm chart from airflorw 1.8.1 version

If I remove securityContext from yaml then it works but i am sure of its impact on application execution.

> Error creating: pods "apache-airflow-webserver-66d7fddcc9-" is forbidden: unable to validate against any security context constraint: [provider "anyuid": Forbidden: not usable by user or serviceaccount, provider restricted: .spec.securityContext.fsGroup: Invalid value: []int64{0}: 0 is not an allowed group, spec.initContainers[0].securityContext.runAsUser: Invalid value: 50000: must be in the ranges: [1001110000, 1001119999], spec.containers[0].securityContext.runAsUser: Invalid value: 50000: must be in the ranges: [1001110000, 1001119999], provider restrictedfsgroup: .spec.securityContext.fsGroup: Invalid value: []int64{0}: 0 is not an allowed group, provider "nonroot": Forbidden: not usable by user or serviceaccount, provider "noobaa": Forbidden: not usable by user or serviceaccount, provider "noobaa-endpoint": Forbidden: not usable by user or serviceaccount, provider "hostmount-anyuid": Forbidden: not usable by user or serviceaccount, provider "machine-api-termination-handler": Forbidden: not usable by user or serviceaccount, provider "hostnetwork": Forbidden: not usable by user or serviceaccount, provider "hostaccess": Forbidden: not usable by user or serviceaccount, provider "rook-ceph": Forbidden: not usable by user or serviceaccount, provider "node-exporter": Forbidden: not usable by user or serviceaccount, provider "privileged": Forbidden: not usable by user or serviceaccount, provider "privileged-genevalogging": Forbidden: not usable by user or serviceaccount, provider "rook-ceph-csi": Forbidden: not usable by user or serviceaccount]

**Reply by potiuk** [reply]
It looks like this is a requirement of the OpenShift not airflow to use the right IDs. Airlfow Helm chart allows you to specify the right ids iun the chart - if your OpenShift requires it. There is a nice section explaining in our docs what you can do with security context: 

https://airflow.apache.org/docs/helm-chart/stable/production-guide.html#security-context

BTW. When you got it to work it would be great if you contribute back some documentation on what needs to be specified to work on OpenShift - this is how it works here - if someone uses a system that has some specific expectations contributing back docs: "And if you have OpenShift" with this and that expectations you should configure it this and that is something tha our users who solve it, often do. For example this is how this part of the documentation has been created:

https://airflow.apache.org/docs/helm-chart/stable/index.html#installing-the-chart-with-argo-cd-flux-or-terraform

(You can click on Suggest a change on this page to open a PR).

I would love if you work it out if you can contribute back such guidelines. That's the least you can do as a user of a free software to have other users :)
